src/DS/sds.c

```diff
@@ -246,108 +270,179 @@ static int ds_sds_dump_component(const char* external_file, const char* componen
+
+	char* mem = oscap_acquire_url_download(url, &memory_size);
+	if (mem == NULL) {
+		ds_sds_session_remote_resources_progress(session)(false, "error", url);
 		return -1;
 	}
 
-	char* xlink_href = (char*)xmlGetNsProp(component_ref, BAD_CAST "href", BAD_CAST xlink_ns_uri);
+	ds_sds_session_remote_resources_progress(session)(false, "ok", url);
+
+	struct oscap_source *source_file = oscap_source_new_take_memory(mem, memory_size, url);
+	xmlDoc *doc = oscap_source_get_xmlDoc(source_file);
+
+	if (doc == NULL) {
+		ret = -1;
+		goto cleanup;
+	}
+
+	xmlNodePtr inner_root = ds_sds_get_component_root_by_id(doc, component_id);
+
+	if (ds_sds_register_component(session, doc, inner_root, component_id, target_filename_dirname, relative_filepath) != 0) {
+		ret = -1;
+		goto cleanup;
+	}
+
+	cleanup:
+		oscap_source_free(source_file);
+		return ret;
+}
+
+static char *compose_target_filename_dirname(const char *relative_filepath, const char* sub_dir)
+{
+	char* filename_cpy = oscap_sprintf("./%s", relative_filepath);
+	char* file_reldir = dirname(filename_cpy);
+
+	char* target_filename_dirname = oscap_sprintf("%s/%s",sub_dir, file_reldir);
+	oscap_free(filename_cpy);
+
+	return target_filename_dirname;
+}
+
+static int ds_sds_dump_component_by_href(struct ds_sds_session *session, char* xlink_href, char *target_filename_dirname, const char* relative_filepath, char* cref_id, char **component_id)
+{
 	if (!xlink_href || strlen(xlink_href) < 2)
 	{
 		oscap_seterr(OSCAP_EFAMILY_XML, "No or invalid xlink:href attribute on given component-ref.");
-		xmlFree(cref_id);
-		xmlFree(xlink_href);
 		return -1;
 	}
 
-
-	const char* component_id;
-	const char* filename;
-
 	if (xlink_href[0] == '#')
 	{
-		filename = NULL;
-		component_id = xlink_href + 1;
+		*component_id = xlink_href + 1;
+
+		return ds_sds_dump_local_component(*component_id, session, target_filename_dirname, relative_filepath);
 
 	} else if (oscap_str_startswith(xlink_href, "file:")){
 
 		char* sep = strchr(xlink_href, '#');
 
-		filename = xlink_href + strlen("file:");
+		const char *filename = xlink_href + strlen("file:");
 
 		if (sep == NULL) {
-			component_id = NULL;
+			*component_id = NULL;
 		} else {
 			*sep = '\0';
-			component_id = sep + 1;
+			*component_id = sep + 1;
 		}
 
+		return ds_sds_dump_file_component(filename, *component_id, session, target_filename_dirname, relative_filepath);
+
+	} else if (oscap_acquire_url_is_supported(xlink_href)){
+		char *sep = strchr(xlink_href, '#');
+
+		char* url = xlink_href;
 
+		if (sep == NULL) {
+			*component_id = NULL;
+		} else {
+			*sep = '\0';
+			*component_id = sep + 1;
+		}
+
+		if (!ds_sds_session_fetch_remote_resources(session)) {
+			static bool fetch_remote_resources_suggested = false;
+
+			if (!fetch_remote_resources_suggested) {
+				fetch_remote_resources_suggested = true;
+				ds_sds_session_remote_resources_progress(session)(false, "Datastream component '%s' points out to the remote '%s'. "
```

Comment:
Would suggest to make it a warning so that scap-workbench does not parse it while reading rule id and results.
... (true, "WARNING: Datastream component ...